# Business Understanding

## Motivation
I just moved back to the United States and now I find myself needing to find a place to live. While price and distance are important to me, safety and crime rates are of the most importance. 

The purpose is to use demographic and crime data to predict crime rates in some of the cities I am interest in moving to.

## Questions to Answer
#####  Can I predict crimes of rape, arson, vehicle theft, larceny, burglary, robbery, and aggravated assault using demographic data?

Crime rates will be the largest factor while I try to determine where to live.

##### Can I determine which demographic fields are the most inflential for such predictions? 

To gain insight into the particular demographics that are of the most influence to the model.

##### Can I apply the models to predict crime rates for some of the cities I'm considering moving to?

I cannot assume that the model will work on cities. If the model does work, I'll follow through and use the results to help me decide.


<hr />

# Data Understanding

## Imports
import the required modules and read in the datasets.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
import seaborn as sns
%matplotlib inline

## Dataset Review

### Crime_df (Crime Features)

The _crime_df_ dataset contains the crimes reported by counties in 2017. We are interested in predicting the crimes:
1. Rape (RAPE)
2. Arson (ARSON)
3. Motor Vehicle Theft (MVTHEFT)
4. Larceny (LARCENY)
5. Burglary (BURGLRY)
6. Robbery (Robbery)
7. Aggravated Assult (AGASSLT)


In [2]:
#source: https://www.kaggle.com/mikejohnsonjr/united-states-crime-rates-by-county
crime_df = pd.read_csv('./crime_data_w_population_and_crime_rate.csv')
crime_df.head(5)

county_name  crime_rate_per_100000  index  EDITION  PART  IDNO  \
0     St. Louis city, MO            1791.995377      1        1     4  1612   
1  Crittenden County, AR            1754.914968      2        1     4   130   
2   Alexander County, IL            1664.700485      3        1     4   604   
3      Kenedy County, TX            1456.310680      4        1     4  2681   
4     De Soto Parish, LA            1447.402430      5        1     4  1137   

   CPOPARST  CPOPCRIM  AG_ARRST  AG_OFF  ...  RAPE  ROBBERY  AGASSLT  BURGLRY  \
0    318667    318667        15      15  ...   200     1778     3609     4995   
1     50717     50717         4       4  ...    38      165      662     1482   
2      8040      8040         2       2  ...     2        5      119       82   
3       444       444         1       1  ...     3        1        2        5   
4     26971     26971         3       3  ...     4       17      368      149   

   LARCENY  MVTHEFT  ARSON  population  FIPS_ST  FIPS_CTY  
0    13791     3543    464      318416       29       510  
1     1753      189     28       49746        5        35  
2      184       12      2        7629       17         3  
3        4        4      0         412       48       261  
4      494       60      0       27083       22        31  

[5 rows x 24 columns]

In [3]:
# Summary Statistics
crime_df.describe()

crime_rate_per_100000        index  EDITION    PART         IDNO  \
count            3136.000000  3136.000000   3136.0  3136.0  3136.000000   
mean              235.085665  1568.500000      1.0     4.0  1590.655612   
std               200.502872   905.429548      0.0     0.0   916.038464   
min                 0.000000     1.000000      1.0     4.0     1.000000   
25%                94.641916   784.750000      1.0     4.0   797.750000   
50%               185.948066  1568.500000      1.0     4.0  1588.500000   
75%               321.367803  2352.250000      1.0     4.0  2385.250000   
max              1791.995377  3136.000000      1.0     4.0  3177.000000   

           CPOPARST      CPOPCRIM     AG_ARRST       AG_OFF       COVIND  ...  \
count  3.136000e+03  3.136000e+03  3136.000000  3136.000000  3136.000000  ...   
mean   9.977578e+04  9.841005e+04     6.201212     5.400510    98.154624  ...   
std    3.198554e+05  3.189017e+05     8.104974     7.532491     8.311058  ...   
min    0.000000e+00  0.000000e+00     0.000000     0.000000     0.000000  ...   
25%    1.120325e+04  1.013975e+04     2.000000     2.000000   100.000000  ...   
50%    2.596550e+04  2.478250e+04     4.000000     3.000000   100.000000  ...   
75%    6.628100e+04  6.603775e+04     7.000000     6.000000   100.000000  ...   
max    9.980757e+06  9.980757e+06   150.000000   140.000000   100.000000  ...   

              RAPE       ROBBERY       AGASSLT       BURGLRY        LARCENY  \
count  3136.000000   3136.000000   3136.000000   3136.000000    3136.000000   
mean     26.324936    112.447704    238.644133    660.654337    1923.397959   
std      79.309098    666.311425    931.091864   2242.359745    6435.993731   
min       0.000000      0.000000      0.000000      0.000000       0.000000   
25%       1.000000      0.000000     10.000000     37.000000      81.000000   
50%       5.000000      4.000000     36.000000    131.000000     348.500000   
75%      19.000000     22.000000    124.000000    416.500000    1205.250000   
max    1976.000000  18923.000000  23057.000000  49107.000000  142604.000000   

            MVTHEFT        ARSON    population      FIPS_ST     FIPS_CTY  
count   3136.000000  3136.000000  3.136000e+03  3136.000000  3136.000000  
mean     227.536671    16.765625  1.007329e+05    30.339286   103.628827  
std     1253.236660    67.970823  3.226566e+05    15.112433   107.987364  
min        0.000000     0.000000  9.000000e+01     1.000000     1.000000  
25%        6.000000     0.000000  1.109275e+04    19.000000    35.000000  
50%       22.000000     3.000000  2.581250e+04    29.000000    79.000000  
75%       74.000000     9.000000  6.758225e+04    45.000000   133.000000  
max    40555.000000  2210.000000  1.001707e+07    56.000000   840.000000  

[8 rows x 23 columns]

In [4]:
crime_df.shape

(3136, 24)

crime_df has 3136 rows and 24 columns.

In [5]:
crime_df.head(2)

county_name  crime_rate_per_100000  index  EDITION  PART  IDNO  \
0     St. Louis city, MO            1791.995377      1        1     4  1612   
1  Crittenden County, AR            1754.914968      2        1     4   130   

   CPOPARST  CPOPCRIM  AG_ARRST  AG_OFF  ...  RAPE  ROBBERY  AGASSLT  BURGLRY  \
0    318667    318667        15      15  ...   200     1778     3609     4995   
1     50717     50717         4       4  ...    38      165      662     1482   

   LARCENY  MVTHEFT  ARSON  population  FIPS_ST  FIPS_CTY  
0    13791     3543    464      318416       29       510  
1     1753      189     28       49746        5        35  

[2 rows x 24 columns]

### Demo_df (Demographic Features)

The demo_df dataset contains US Census data that will be used as additional features to try to predict crime with.

In [6]:
#source: https://hub.arcgis.com/datasets/6afda9afb9f94e27adf39d36e119eb29_2
demo_df= pd.read_csv('./Census_demographics_by_County.csv')
demo_df.head(5)

OBJECTID  STATE  STCN        CNTYNAME    AREALAND  AREAWATER  TOTALPOP  \
0         1      1  1001  Autauga County  1539584444   25773561     55136   
1         2      1  1005  Barbour County  2291820953   50864677     27119   
2         3      1  1023  Choctaw County  2365954803   19059247     13546   
3         4      1  1033  Colbert County  1534878355   80029923     54491   
4         5      1  1047   Dallas County  2534807336   39134779     42743   

     POP_DEN  NUM_MINORITY  PCT_MINORITY  ...  EMP_STAT_NOT_IN_LF  \
0  92.753326         13103     23.764872  ...               15561   
1  30.647195         14614     53.888418  ...               11086   
2  14.828676          5995     44.256607  ...                5594   
3  91.949330         11392     20.906205  ...               20519   
4  43.673480         30492     71.337997  ...               15660   

   PCT_EMP_STAT_NOT_IN_LF  Shape_Length  Shape_Area  FBPOP  EUROPE   ASIA  \
0               36.603782      2.052620    0.150260  772.0   141.0  150.0   
1               50.736842      2.566611    0.223256  769.0    14.0  169.0   
2               50.785293      2.684498    0.227716   30.0    12.0   16.0   
3               46.439888      2.174954    0.158871  997.0    51.0  197.0   
4               47.866487      3.008813    0.246517  252.0    89.0   24.0   

   AFRICA  OCEANIA  AMERICAS  
0     0.0    119.0     362.0  
1     0.0      0.0     586.0  
2     0.0      0.0       2.0  
3     0.0      0.0     749.0  
4    14.0      0.0     125.0  

[5 rows x 266 columns]

In [7]:
demo_df.describe()

OBJECTID        STATE          STCN      AREALAND     AREAWATER  \
count  3220.00000  3220.000000   3220.000000  3.220000e+03  3.220000e+03   
mean   1610.50000    31.290683  31393.643478  2.843643e+09  2.145640e+08   
std     929.67826    16.277207  16292.019724  9.250791e+09  1.213797e+09   
min       1.00000     1.000000   1001.000000  5.146977e+06  0.000000e+00   
25%     805.75000    19.000000  19032.500000  1.085116e+09  7.030551e+06   
50%    1610.50000    30.000000  30024.000000  1.565591e+09  1.949135e+07   
75%    2415.25000    46.000000  46107.500000  2.368926e+09  6.111069e+07   
max    3220.00000    72.000000  72153.000000  3.768559e+11  2.519064e+10   

           TOTALPOP       POP_DEN  NUM_MINORITY  PCT_MINORITY         WHITE  \
count  3.220000e+03   3220.000000  3.220000e+03   3220.000000  3.220000e+03   
mean   9.867890e+04    287.536717  3.744130e+04     24.278493  7.278984e+04   
std    3.162316e+05   1769.390838  1.946284e+05     22.941132  1.985159e+05   
min    7.300000e+01      0.038837  4.000000e+00      0.177148  3.200000e+01   
25%    1.123600e+04     17.424052  1.134750e+03      6.482221  9.264250e+03   
50%    2.608800e+04     46.566340  4.330500e+03     15.663647  2.198650e+04   
75%    6.618100e+04    130.364554  1.546775e+04     35.342812  5.619225e+04   
max    9.974203e+06  71434.087691  7.262538e+06    100.000000  5.329333e+06   

       ...  EMP_STAT_NOT_IN_LF  PCT_EMP_STAT_NOT_IN_LF  Shape_Length  \
count  ...        3.220000e+03             3220.000000   3220.000000   
mean   ...        2.838379e+04               40.913971      2.580587   
std    ...        8.663136e+04                8.175617      3.553004   
min    ...        1.300000e+01                9.888626      0.111575   
25%    ...        3.949500e+03               35.001246      1.622222   
50%    ...        8.945000e+03               40.090688      1.973096   
75%    ...        2.137700e+04               45.845911      2.594886   
max    ...        2.796585e+06               79.413004     71.098879   

        Shape_Area         FBPOP         EUROPE          ASIA        AFRICA  \
count  3220.000000  3.142000e+03    3142.000000  3.142000e+03   3142.000000   
mean      0.342839  1.306698e+04    1531.975175  3.818908e+03    554.621260   
std       1.789178  8.951773e+04    8654.639398  2.963884e+04   3020.409952   
min       0.000536  0.000000e+00       0.000000  0.000000e+00      0.000000   
25%       0.114594  1.720000e+02      20.000000  2.400000e+01      0.000000   
50%       0.164876  6.215000e+02      81.000000  1.040000e+02      8.000000   
75%       0.246606  2.719750e+03     367.500000  5.585000e+02     75.000000   
max      74.225434  3.484323e+06  254163.000000  1.191920e+06  54798.000000   

            OCEANIA      AMERICAS  
count   3142.000000  3.142000e+03  
mean      74.039465  7.087434e+03  
std      570.714697  5.431783e+04  
min        0.000000  0.000000e+00  
25%        0.000000  7.700000e+01  
50%        0.000000  3.320000e+02  
75%       16.000000  1.582750e+03  
max    17400.000000  2.051695e+06  

[8 rows x 265 columns]

In [8]:
demo_df.shape

(3220, 266)

demo_df has 3220 rows and 266 columns. 

__NOTE:__ Some rows and columns on demo_df and crime_df will not be joined due to the row difference,

### Demo_state_df (used for data preparation)

The demo_state_df dataset is used only for data.

In [9]:
#source: https://hub.arcgis.com/datasets/utah::us-states
demo_state_df = pd.read_csv("./US_States.csv")
demo_state_df.head(5)

OBJECTID_1  OBJECTID  STATE_NAME  STATE_FIPS STATE_ABBR  SHAPE_Length  \
0           1       NaN     Alabama           1         AL     20.554796   
1           2       NaN      Alaska           2         AK    318.661159   
2           3       NaN     Arizona           4         AZ     23.916224   
3           4       NaN    Arkansas           5         AR     22.058710   
4           5       NaN  California           6         CA     55.520183   

   SHAPE_Area  
0   12.881818  
1  167.784052  
2   28.921383  
3   13.585572  
4   41.639766

In [10]:
demo_state_df.describe()

OBJECTID_1  OBJECTID  STATE_FIPS  SHAPE_Length  SHAPE_Area
count   51.000000       0.0   51.000000     51.000000   51.000000
mean    26.000000       NaN   28.960784     36.852386   19.354421
std     14.866069       NaN   15.832828     45.643270   24.622024
min      1.000000       NaN    1.000000      0.672896    0.018363
25%     13.500000       NaN   16.500000     20.647511    9.772642
50%     26.000000       NaN   29.000000     23.860175   13.892748
75%     38.500000       NaN   41.500000     36.245103   22.776599
max     51.000000       NaN   56.000000    318.661159  167.784052

In [11]:
demo_state_df.shape

(51, 7)

<hr />

# Data Preparation

### Add Explicit State Abbreviation to Demo_df
Start by adding a state abbreviation field to the demo_df dataset by matching on the numeric state designator shared by demo_df and demo_state_df

In [12]:
#Use demo_state_df STATE_FIPS column to determine the state in demo_df

#Match on the state numeric code
for name in range(len(demo_df['STATE'])):
    for stateCode in range(len(demo_state_df['STATE_FIPS'])):
        if(demo_df['STATE'][name] == demo_state_df['STATE_FIPS'][stateCode]):
            demo_df['STATE'][name] = demo_state_df['STATE_ABBR'][stateCode]

<ipython-input-12-d66b2ea9bf7b>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo_df['STATE'][name] = demo_state_df['STATE_ABBR'][stateCode]
/home/ada/pythonJupyter/jupyterenvironment/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


### Split County Name and State Abbreviation

The county name and the state abbreviation are listed together as a string in the crime_df. Mutate field from a string to a list of strings by splitting on the comma. This will help for matching on the state later.

In [13]:
#Split string on comma and turn into a list. Needed for matching later.
for name in range(len(crime_df['county_name'])):
    crime_df['county_name'][name] = crime_df['county_name'][name].split(",")
    crime_df['county_name'][name][1] = crime_df['county_name'][name][1].strip()

<ipython-input-13-3bc2aae1c169>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crime_df['county_name'][name] = crime_df['county_name'][name].split(",")


### Add a dummy join column
Add a join column to keep record of rows that refer to the same county in the crime_df and the demo_df

In [14]:
# Add columns for matching
crime_df.insert(0,"JoinColumn", [np.nan] * crime_df.shape[0], True )
demo_df.insert(0,"JoinColumn", [np.nan] * demo_df.shape[0], True )

### Match crime_df and demo_df
Match the crime_df and the demo_df on both county and state, as states can have counties with the same name.

In [15]:
# Match based on the county name and the state
numEqual = 0
for x in range(len(crime_df['county_name'])):
    for y in range(len(demo_df['CNTYNAME'])):
        if crime_df['county_name'][x][0] == demo_df['CNTYNAME'][y]:
            if(crime_df['county_name'][x][1] == demo_df['STATE'][y]):
                numEqual += 1
                crime_df['JoinColumn'][x] = numEqual
                demo_df['JoinColumn'][y] = numEqual

<ipython-input-15-bb558d746821>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crime_df['JoinColumn'][x] = numEqual
<ipython-input-15-bb558d746821>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo_df['JoinColumn'][y] = numEqual


#### Matching result

The prints below show that we have 3135 matches and had to drop a little under 100 counties from the demo_df dataset. 

##### Note on data dropping
The counties that did not have a match have to be dropped because either crime or demographic data is missing. These counties just are not in both datasets.

One county was dropped from the crime_df dataset but 85 were dropped from demo_df. This means we were missing crime data for 85 counties.

Both the demo_df and crime_df have a unique county on each row, but demo_df has 84 more counties that crime_df, consequently, not all counties in demo_df can be mapped to crime in crime_df

In [16]:
#Show the number of matches                
print(numEqual)
print(crime_df.shape)
print(demo_df.shape)

#Drop Columns that did not have a match
crime_df_dropped = crime_df.dropna(subset=['JoinColumn'], how = 'any')
demo_df_dropped = demo_df.dropna(subset=['JoinColumn'], how = 'any')

#Show the number of matches in each data frame
#Note that they are the same size, there were not any duplicate matches
print(crime_df_dropped.shape)
print(demo_df_dropped.shape)

3135
(3136, 25)
(3220, 267)
(3135, 25)
(3135, 267)


### Aggregating Data and Imputation

Next, join the crime_df and demo_df datasets based on the matching column.

Impute values for the cities:
1. Columbia, Howard County
2. Odenton, Anne Arundel County
3. Glen Burnie, Anne Arundel County
4. Hanover, Anne Arundel County

Then drop the identification based fields.

#### On Imputation Methodology

I could not find a dataset for the cities that consisted of the same fields found in the demographic demo_df dataset. So I extract the rows for the counties the cites are located in then reduce the numeric fields by the difference in population. For example, if Howard County had 1000 Asian females and Columbia had 1/4 of the population of Howard County, then the value of the field for Asian females would be 250.


It basically assumes that the counties have a uniform demographic distribution. While that is is not a good assumption, the result is for personal estimation only, so I accept the inaccuracies associated with the assumption.

In [17]:

#Columns that are not to be used in the modeling and will be dropped.
dropList = ['JoinColumn','county_name', 'EDITION', 'PART', 'IDNO', 'INDEX', 'FIPS_ST', 'FIPS_CTY','OBJECTID', 'STATE', 'STCN', 'CNTYNAME']

#Join on the matches
joinedDF = crime_df_dropped.merge(demo_df_dropped, on='JoinColumn')

#Subset by county based on potential living city locations
columbia = joinedDF[joinedDF['STATE'] =='MD'].copy()
columbia = columbia[columbia['CNTYNAME'] == "Howard County"]

odenton = joinedDF[joinedDF['STATE'] =='MD'].copy()
odenton = odenton[odenton['CNTYNAME'] == "Anne Arundel County"]

glen_burnie = joinedDF[joinedDF['STATE'] =='MD'].copy()
glen_burnie = glen_burnie[glen_burnie['CNTYNAME'] == "Anne Arundel County"]

hanover = joinedDF[joinedDF['STATE'] =='MD'].copy()
hanover = hanover[hanover['CNTYNAME'] == "Anne Arundel County"]

#Drop unneccessary columns
columbia = columbia.drop(dropList, axis =1)
odenton = odenton.drop(dropList, axis =1)
glen_burnie = glen_burnie.drop(dropList, axis =1)
hanover = hanover.drop(dropList, axis =1)
joinedDF = joinedDF.drop(dropList, axis =1)

# Population in 2019
# Used for imputation
howardCountyPop = 325690
columbiaPop = 103991
anneArundelPop = 579234
odentonPop = 41846
glen_burniePop = 69872
hanoverPop = 12952

#Impute the numeric fields based on the population proportion
def imputeSubset(df, portion):
    """
    Description: This function iterates through the columns of a dataframe
    and finds columns that are numeric. It then reduces or increases the values 
    of that column by the portion provided.

    Arguments:
        df: - the dataframe that is to be mutated 
        portion: - the amount to increase or decrease each value by.

    Returns:
        None: The passed dataframe is mutated.
    """
    for column in df.columns:
        if np.issubdtype(df.dtypes[column], np.number):
                df[column] = df[column] * portion

                
#Impute values for the cities from their county
imputeSubset(columbia, columbiaPop / howardCountyPop)
imputeSubset(odenton, odentonPop / anneArundelPop)
imputeSubset(glen_burnie, glen_burniePop / anneArundelPop)
imputeSubset(hanover, hanoverPop / anneArundelPop)

<hr />

# Data Modeling

The function defined below is used to create linear regression models of:

1. Rape (RAPE)
2. Arson (ARSON)
3. Motor Vehicle Theft (MVTHEFT)
4. Larceny (LARCENY)
5. Burglary (BURGLRY)
6. Robbery (Robbery)
7. Aggravated Assult (AGASSLT)

for the cities: 

1. Columbia, Howard County
2. Odenton, Anne Arundel County
3. Glen Burnie, Anne Arundel County
4. Hanover, Anne Arundel County

#### Be sure to note the comment on imputation for the cities from the county data set

In [18]:
def model(parameter, joinedDF):
    """
    Description: The function performs a linear regression on the passed data frame
    to predict the passed parameter. Note: A lot of the methodoloy here came from what
    I learned in the Udacity Data Science track.
    
    Arguments:
        parameter: The parameter to predict
        joinedDF: A cleaned dataframe that should contain the parameter of interest.
        
    Returns:
        List: A list containing the predictions for the cities of interest, the coeffiecents from
              the model, and a dataframe comparing the actual and predicted values.
    """
    X = joinedDF.drop([parameter], axis = 1)
    y = joinedDF[parameter]
    #Split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .30, random_state=101011)
    # train
    lm_model = LinearRegression(normalize=True) # Instantiate
    lm_model.fit(X_train, y_train) #Fit
    # predict
    y_test_preds = lm_model.predict(X_test)
    y_train_preds = lm_model.predict(X_train)
    #score
    test_score = r2_score(y_test, y_test_preds)
    train_score = r2_score(y_train, y_train_preds)
    print("The rsquared on the training data was {}.  The rsquared on the test data was {}.".format(train_score, test_score))
    
    #For review of the model's performance
    coefs_df = pd.DataFrame()
    coefs_df['est_int'] = X_train.columns
    coefs_df['coefs'] = lm_model.coef_
    coefs_df['abs_coefs'] = np.abs(lm_model.coef_)
    coefs_df = coefs_df.sort_values('abs_coefs', ascending=False)
    compareDF = pd.DataFrame({'Actual': y_test, 'Predicted': y_test_preds})
    
    #Subset the city data based on the passed parameter
    columbiaDropped = columbia.drop([parameter], axis = 1)
    odentonDropped =odenton.drop([parameter], axis = 1)
    glen_burnieDropped = glen_burnie.drop([parameter], axis = 1)
    hanoverDropped = hanover.drop([parameter], axis = 1)
    
    #Attempt to predict the parameter using the model
    columbiaVal = lm_model.predict(columbiaDropped)[0] 
    odentonVal = lm_model.predict(odentonDropped)[0] 
    glen_burnieVal = lm_model.predict(glen_burnieDropped)[0] 
    hanoverVal = lm_model.predict(hanoverDropped)[0] 
    
    return [[columbiaVal, odentonVal,glen_burnieVal,hanoverVal], coefs_df, compareDF] 

<hr />

# Results

## Executing The Linear Regression

The following cells call the model function on the parameter of interest then print the __rsquared score__, the most __impactful fields__, and a __comparison of the predicted and actual values__

In [19]:
results = pd.DataFrame({'City' : ["Columbia", "Odenton", "Glen Burnie", "Hanover"]})
results.set_index('City')
print(results)

          City
0     Columbia
1      Odenton
2  Glen Burnie
3      Hanover


In [20]:
result = model("RAPE", joinedDF)
results.insert(1, "RAPE",result[0])
result[2]

The rsquared on the training data was 0.9737776942603972.  The rsquared on the test data was 0.7162012091176642.


Actual   Predicted
2524       0   -0.218750
3102       0  -10.609375
82        31   26.984375
1573      94  109.734375
2632       1   -0.343750
...      ...         ...
2501       7    1.562500
966        8    7.906250
439      289  318.546875
2943       0   -6.781250
2835       2    6.593750

[941 rows x 2 columns]

In [21]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  1.251427e+12  1.251427e+12
259      PCT_EMP_STAT_IN_LF  1.251427e+12  1.251427e+12
171           PCT_HLI_IE_LI  1.507803e+11  1.507803e+11
173          PCT_HLI_IE_NLI  9.297506e+10  9.297506e+10
59             PCT_AGE_GT17 -8.781943e+10  8.781943e+10
53             PCT_AGE_LT18 -8.781943e+10  8.781943e+10
169              PCT_HLI_IE -8.753956e+10  8.753956e+10
181           PCT_HLI_OTHER  8.232970e+10  8.232970e+10
185       PCT_HLI_OTHER_NLI -7.689419e+10  7.689419e+10
148           PCT_LAN_ENG_W -6.921693e+10  6.921693e+10
150          PCT_LAN_ENG_NW -5.860229e+10  5.860229e+10
152          PCT_LAN_ENG_NA -5.860229e+10  5.860229e+10
159             PCT_LINGISO -5.780523e+10  5.780523e+10
175             PCT_HLI_API  5.022252e+10  5.022252e+10
179         PCT_HLI_API_NLI -4.478702e+10  4.478702e+10
156            PCT_ENG_LTVW  4.334871e+10  4.334871e+10
144         PCT_LAN_NON_ENG  3.643051e+10  3.643051e+10
165      PCT_HLI_SPANISH_LI  3.236567e+10  3.236567e+10
163         PCT_HLI_SPANISH  3.087507e+10  3.087507e+10
21             PCT_MINORITY -2.851222e+10  2.851222e+10

In [22]:
result = model("ARSON", joinedDF)
results.insert(1, "ARSON",result[0])
result[2]
#Predictions for ARSON are very inaccuarate

The rsquared on the training data was 0.9539673580763047.  The rsquared on the test data was 0.3397538395428641.


Actual  Predicted
2524       0   -2.40625
3102       0   -3.37500
82        13   11.84375
1573      39   94.84375
2632       1   -1.50000
...      ...        ...
2501       3    0.78125
966        1    8.06250
439      169  150.37500
2943       0   -4.56250
2835       5    5.87500

[941 rows x 2 columns]

In [23]:
result[1].head(20)

est_int         coefs     abs_coefs
259      PCT_EMP_STAT_IN_LF -1.740746e+12  1.740746e+12
269  PCT_EMP_STAT_NOT_IN_LF -1.740746e+12  1.740746e+12
179         PCT_HLI_API_NLI  2.412376e+11  2.412376e+11
175             PCT_HLI_API -2.234421e+11  2.234421e+11
171           PCT_HLI_IE_LI -1.938352e+11  1.938352e+11
183        PCT_HLI_OTHER_LI -1.578975e+11  1.578975e+11
177          PCT_HLI_API_LI  1.282647e+11  1.282647e+11
159             PCT_LINGISO  1.129729e+11  1.129729e+11
169              PCT_HLI_IE  9.865786e+10  9.865786e+10
173          PCT_HLI_IE_NLI -8.086235e+10  8.086235e+10
165      PCT_HLI_SPANISH_LI -7.285435e+10  7.285435e+10
148           PCT_LAN_ENG_W  7.136630e+10  7.136630e+10
181           PCT_HLI_OTHER  6.272007e+10  6.272007e+10
150          PCT_LAN_ENG_NW  5.691490e+10  5.691490e+10
152          PCT_LAN_ENG_NA  5.691490e+10  5.691490e+10
185       PCT_HLI_OTHER_NLI -4.492456e+10  4.492456e+10
156            PCT_ENG_LTVW -4.013858e+10  4.013858e+10
167     PCT_HLI_SPANISH_NLI  4.011854e+10  4.011854e+10
144         PCT_LAN_NON_ENG -3.311872e+10  3.311872e+10
146          PCT_LAN_ENG_VW  3.122772e+10  3.122772e+10

In [24]:
result = model("MVTHEFT", joinedDF)
results.insert(1, "MVTHEFT",result[0])
result[2]

The rsquared on the training data was 0.9999999521064714.  The rsquared on the test data was 0.9999998983422953.


Actual    Predicted
2524       4     4.035400
3102       0     0.294189
82       119   118.950195
1573     464   464.180664
2632       0    -0.067383
...      ...          ...
2501      14    14.032715
966       34    33.833008
439     3967  3967.805664
2943       1     0.883057
2835      31    30.875000

[941 rows x 2 columns]

In [25]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  9.995406e+09  9.995406e+09
259      PCT_EMP_STAT_IN_LF  9.995406e+09  9.995406e+09
177          PCT_HLI_API_LI -3.694731e+09  3.694731e+09
179         PCT_HLI_API_NLI -3.188915e+09  3.188915e+09
175             PCT_HLI_API  3.086010e+09  3.086010e+09
59             PCT_AGE_GT17  2.967552e+09  2.967552e+09
53             PCT_AGE_LT18  2.967552e+09  2.967552e+09
183        PCT_HLI_OTHER_LI -1.833522e+09  1.833522e+09
169              PCT_HLI_IE -1.369639e+09  1.369639e+09
152          PCT_LAN_ENG_NA  1.360300e+09  1.360300e+09
150          PCT_LAN_ENG_NW  1.360300e+09  1.360300e+09
185       PCT_HLI_OTHER_NLI -1.327706e+09  1.327706e+09
173          PCT_HLI_IE_NLI  1.266735e+09  1.266735e+09
181           PCT_HLI_OTHER  1.224802e+09  1.224802e+09
63              PCT_FEMALES  1.055704e+09  1.055704e+09
61                PCT_MALES  1.055704e+09  1.055704e+09
154         PCT_LAN_ENG_LTW -1.034292e+09  1.034292e+09
171           PCT_HLI_IE_LI  7.609186e+08  7.609186e+08
146          PCT_LAN_ENG_VW  7.530075e+08  7.530075e+08
261  PCT_EMP_STAT_IN_CIV_LF  6.418708e+08  6.418708e+08

In [26]:
result = model("LARCENY", joinedDF)
results.insert(1, "LARCENY",result[0])
result[2]

The rsquared on the training data was 0.9999999979573293.  The rsquared on the test data was 0.999999997083696.


Actual     Predicted
2524      21     21.036621
3102       2      2.278809
82      1600   1599.951172
1573    7010   7010.199463
2632      42     41.932373
...      ...           ...
2501     169    169.031494
966      610    609.831055
439    20856  20856.776123
2943      17     16.878906
2835     804    803.872559

[941 rows x 2 columns]

In [27]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  1.021766e+10  1.021766e+10
259      PCT_EMP_STAT_IN_LF  1.021766e+10  1.021766e+10
177          PCT_HLI_API_LI -3.750320e+09  3.750320e+09
179         PCT_HLI_API_NLI -3.239929e+09  3.239929e+09
175             PCT_HLI_API  3.136848e+09  3.136848e+09
59             PCT_AGE_GT17  2.973672e+09  2.973672e+09
53             PCT_AGE_LT18  2.973672e+09  2.973672e+09
183        PCT_HLI_OTHER_LI -1.840216e+09  1.840216e+09
169              PCT_HLI_IE -1.367656e+09  1.367656e+09
152          PCT_LAN_ENG_NA  1.353940e+09  1.353940e+09
150          PCT_LAN_ENG_NW  1.353940e+09  1.353940e+09
185       PCT_HLI_OTHER_NLI -1.329825e+09  1.329825e+09
173          PCT_HLI_IE_NLI  1.264575e+09  1.264575e+09
181           PCT_HLI_OTHER  1.226745e+09  1.226745e+09
154         PCT_LAN_ENG_LTW -1.034873e+09  1.034873e+09
171           PCT_HLI_IE_LI  7.541841e+08  7.541841e+08
146          PCT_LAN_ENG_VW  7.513078e+08  7.513078e+08
261  PCT_EMP_STAT_IN_CIV_LF  6.414967e+08  6.414967e+08
267      PCT_EMP_STAT_IN_AF  6.414967e+08  6.414967e+08
165      PCT_HLI_SPANISH_LI -5.669825e+08  5.669825e+08

In [28]:
result = model("BURGLRY", joinedDF)
results.insert(1, "BURGLRY",result[0])
result[2]

The rsquared on the training data was 0.9999999816071707.  The rsquared on the test data was 0.9999999793052033.


Actual    Predicted
2524       4     4.033447
3102       0     0.291016
82       734   733.950928
1573    2155  2155.163330
2632      13    12.932373
...      ...          ...
2501     100   100.031738
966      235   234.832275
439     9180  9180.809326
2943       7     6.881104
2835     265   264.871094

[941 rows x 2 columns]

In [29]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF  1.030303e+10  1.030303e+10
259      PCT_EMP_STAT_IN_LF  1.030303e+10  1.030303e+10
177          PCT_HLI_API_LI -3.760505e+09  3.760505e+09
179         PCT_HLI_API_NLI -3.250798e+09  3.250798e+09
175             PCT_HLI_API  3.147183e+09  3.147183e+09
59             PCT_AGE_GT17  2.969849e+09  2.969849e+09
53             PCT_AGE_LT18  2.969849e+09  2.969849e+09
61                PCT_MALES -1.885949e+09  1.885949e+09
63              PCT_FEMALES -1.885949e+09  1.885949e+09
183        PCT_HLI_OTHER_LI -1.833829e+09  1.833829e+09
169              PCT_HLI_IE -1.360925e+09  1.360925e+09
152          PCT_LAN_ENG_NA  1.352245e+09  1.352245e+09
150          PCT_LAN_ENG_NW  1.352245e+09  1.352245e+09
185       PCT_HLI_OTHER_NLI -1.324122e+09  1.324122e+09
173          PCT_HLI_IE_NLI  1.257310e+09  1.257310e+09
181           PCT_HLI_OTHER  1.220508e+09  1.220508e+09
154         PCT_LAN_ENG_LTW -1.033407e+09  1.033407e+09
146          PCT_LAN_ENG_VW  7.508722e+08  7.508722e+08
171           PCT_HLI_IE_LI  7.476034e+08  7.476034e+08
261  PCT_EMP_STAT_IN_CIV_LF  6.400322e+08  6.400322e+08

In [30]:
result = model("AGASSLT", joinedDF)
results.insert(1, "AGASSLT",result[0])
result[2]


The rsquared on the training data was 0.9858276515273506.  The rsquared on the test data was 0.7630907938511257.


Actual  Predicted
2524       2    -74.875
3102       0    128.625
82       394    412.875
1573     895   1062.500
2632       8     11.500
...      ...        ...
2501       7      9.250
966       52     38.500
439     2274    966.375
2943       1     51.375
2835      36     11.750

[941 rows x 2 columns]

In [31]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF -6.239436e+12  6.239436e+12
259      PCT_EMP_STAT_IN_LF -6.239436e+12  6.239436e+12
59             PCT_AGE_GT17 -1.605918e+12  1.605918e+12
53             PCT_AGE_LT18 -1.605918e+12  1.605918e+12
183        PCT_HLI_OTHER_LI  1.012390e+12  1.012390e+12
63              PCT_FEMALES -9.243892e+11  9.243892e+11
61                PCT_MALES -9.243892e+11  9.243892e+11
185       PCT_HLI_OTHER_NLI  7.759634e+11  7.759634e+11
181           PCT_HLI_OTHER -6.766975e+11  6.766975e+11
169              PCT_HLI_IE  5.823059e+11  5.823059e+11
173          PCT_HLI_IE_NLI -4.830400e+11  4.830400e+11
177          PCT_HLI_API_LI  3.845554e+11  3.845554e+11
267      PCT_EMP_STAT_IN_AF  3.350314e+11  3.350314e+11
261  PCT_EMP_STAT_IN_CIV_LF  3.350314e+11  3.350314e+11
154         PCT_LAN_ENG_LTW -3.280026e+11  3.280026e+11
152          PCT_LAN_ENG_NA  3.274594e+11  3.274594e+11
150          PCT_LAN_ENG_NW  3.274594e+11  3.274594e+11
163         PCT_HLI_SPANISH  3.242760e+11  3.242760e+11
171           PCT_HLI_IE_LI -2.466134e+11  2.466134e+11
159             PCT_LINGISO -2.364266e+11  2.364266e+11

In [32]:
result = model("ROBBERY", joinedDF )
results.insert(1, "ROBBERY",result[0])
result[2]

The rsquared on the training data was 0.9969916270011239.  The rsquared on the test data was 0.823755330025302.


Actual    Predicted
2524       1   -12.359375
3102       0    51.523438
82        49    72.218750
1573     280   719.968750
2632       1    -7.226562
...      ...          ...
2501       1    14.585938
966       14    21.843750
439      910  1158.609375
2943       1    19.203125
2835       5    -5.460938

[941 rows x 2 columns]

In [33]:
result[1].head(20)

est_int         coefs     abs_coefs
269  PCT_EMP_STAT_NOT_IN_LF -1.141678e+12  1.141678e+12
259      PCT_EMP_STAT_IN_LF -1.141678e+12  1.141678e+12
177          PCT_HLI_API_LI -4.676747e+11  4.676747e+11
171           PCT_HLI_IE_LI -3.980697e+11  3.980697e+11
183        PCT_HLI_OTHER_LI -3.858850e+11  3.858850e+11
61                PCT_MALES  3.755275e+11  3.755275e+11
63              PCT_FEMALES  3.755275e+11  3.755275e+11
165      PCT_HLI_SPANISH_LI -3.138486e+11  3.138486e+11
159             PCT_LINGISO  2.749814e+11  2.749814e+11
154         PCT_LAN_ENG_LTW  2.439053e+11  2.439053e+11
179         PCT_HLI_API_NLI -1.926933e+11  1.926933e+11
175             PCT_HLI_API  1.757614e+11  1.757614e+11
148           PCT_LAN_ENG_W  1.368417e+11  1.368417e+11
156            PCT_ENG_LTVW -1.330406e+11  1.330406e+11
53             PCT_AGE_LT18  1.274358e+11  1.274358e+11
59             PCT_AGE_GT17  1.274358e+11  1.274358e+11
173          PCT_HLI_IE_NLI -1.230883e+11  1.230883e+11
185       PCT_HLI_OTHER_NLI -1.109036e+11  1.109036e+11
152          PCT_LAN_ENG_NA -1.070636e+11  1.070636e+11
150          PCT_LAN_ENG_NW -1.070636e+11  1.070636e+11

In [34]:
results.sum(axis = 1)

0    6.612362e+14
1    9.012209e+14
2    8.542202e+14
3    9.496773e+14
dtype: float64

## Evaluation And Conclusion

#### Model Evaluation
The results for Arson has the lowest R squared score of 0.3397538395428641, which is too low to comfortably use conclusively. Additionally, Burglary, Motor Vehicle Theft, and Larceny had very high scores. They are so high that I'm concerned I overfitted.

The remaining scores were:
1. Rape: 0.7162012091176642
2. Aggravated Assult: 0.7630907938511257
3. Robbery: 0.823755330025302

So I will use the rape, aggravated assult, and robbery fields to help me decide where I should live.

#### Influential Demographic Fields 

The following fields seemed to have the largest influence on the model:
1. PCT_EMP_STAT_NOT_IN_LF (those not employed in the labor force?)
2. PCT_EMP_STAT_IN_LF	(those who are employed in the labor force?)
3. PCT_HLI_API_LI (Asian pacific Islanders who are linguistically isolated?)
4. PCT_HLI_IE_LI (no clue)
5. PCT_AGE_LT18 (Less than 18 years old)
6. PCT_AGE_GT17 (Greater than 17 years old?)

I stuggled to find documentation on these fields and the phrases in parentheses are the best descriptions I could come up with. 

__NOTE:__ _These fields are used for prediction and do not neccessarily mean causation_. I am not saying that non-english speaking people are the cause of crime in this county

In [35]:
violent = results.drop(['BURGLRY', 'LARCENY', 'MVTHEFT', 'ARSON'], axis = 1)
#Sum for 'RAPE', 'AGASSLT', 'ROBBERY'
violent.sum(axis = 1)

0    5.443797e+14
1    7.419532e+14
2    7.032587e+14
3    7.818462e+14
dtype: float64

In [36]:
prop = results.drop(['RAPE', 'AGASSLT', 'ROBBERY', 'ARSON'], axis = 1)
#Sum for 'BURGLRY', 'LARCENY', 'MVTHEFT'
prop.sum(axis = 1)

0   -2.883787e+12
1   -3.930409e+12
2   -3.725429e+12
3   -4.141737e+12
dtype: float64

In [37]:
all = results.drop(['ARSON'], axis = 1)
all.sum(axis = 1)

0    5.414959e+14
1    7.380228e+14
2    6.995332e+14
3    7.777044e+14
dtype: float64

### Closing Remarks
For the three cities in Anne Arundel County, it seems that the smaller the city, the higher the crime?

This is not mirrored by online sources and is probably indicative that my method of imputation is a lot worse than i had anticipated. It could also be the case that I cannot rely on crime trends in counties to reliable predict crime trends in cities. 